### DISCUSSION TITLE
Discover Cmd

### DISCUSSION THREAD

**Discussion by matijazigic** [ifm/ifm3d#433] [Q&A]
Is it possible to discover all ifm device on network without providing option with exact ip of the device? 
Currently, when I'm trying to find cameras/VPU, it says:

```
2023-12-07 13:48:09 ERROR [../modules/device/src/libifm3d_device/xmlrpc_wrapper.hpp:98] http://192.168.0.69:80/api/rpc/v1/com.ifm.efector/ -> getParameter: Unable to transport XML to server and get XML response back.  libcurl failed to execute the HTTP POST transaction, explaining:  Operation timed out after 3004 milliseconds with 0 bytes received
2023-12-07 13:48:09 WARN [../modules/device/src/libifm3d_device/device.cpp:176] Could not probe device type: Lib: XMLRPC Timeout - can you `ping' the sensor?
ifm3d error: -100001
Lib: XMLRPC Timeout - can you `ping' the sensor
```

It is okay, because it looks only for default IP (192.168.0.69), but what if I don't know exact IP? Should I iterate over all IPs in the subnet, or there is an option already for that?

Another thing is that when I know the address of the device and provide it within 'discover' cmd (`ifm3d discover --ip 192.168.100.191`), I got: `Info: No devices available`
From code I can normally get the device while providing that IP.

Is there any python api for discover cmd? I was looking in the documentation, but cannot find any, so I suppose there is no api for that.

**Comment by graugans** [comment]
There is a way we call discovery. Unfortunately there is a limitation in Linux that a broadcast per interface can only be sent by the root user. Thus this very old project of mine requires root access: https://github.com/ifm/python-ipv4-discovery

The command `ifm3d discover` tries to workaround the root requirement but it does not work in all cases. But  @inbangsa has all the details.

**Reply by graugans** [reply]
I have to check if I can provide a patch to get this old python code to be run on a new Python version

**Reply by graugans** [reply]
```diff
diff --git a/discovery.py b/discovery.py
index 1e91b64..24ee197 100755
--- a/discovery.py
+++ b/discovery.py
@@ -66,7 +66,7 @@ if __name__ == "__main__":
     s.bind(('', PORT))
     s.settimeout(5)

-    print "Trying to find server..."
+    print("Trying to find server...")
     s.sendto(broadcast, ("<broadcast>", PORT))
     try:
         response,server = s.recvfrom(8) # we typically receive our own broadcast
@@ -102,6 +102,6 @@ if __name__ == "__main__":
             # was successful. The easiest way might be to use the broadcast again

     except socket.timeout as e:
-        print "Could not find ipv4 discovery server {}".format(str(e))
+        print("Could not find ipv4 discovery server {}".format(str(e)))
     # closing the socket
     s.close()
````
